### ISSUE TITLE
[LAB] Week 2_Problem 1_송윤근

### ISSUE THREAD

**Issue by youngandtherich** [vlab-kaist/NN101_23S#101]
import torch
from random import random 
from typing import Callable

import numpy as np
##                                                            ##

# NOTE : Feel free to use torch.optim and tensor.

def training(x_train : list[float], y_train : list[float]) -> list[float]: # DO NOT MODIFY FUNCTION NAME
    # Data normalization code (prevents overflow when calculating MSE, prevents underfitting)
    # Note that you need to convert [w, b] to the original scale before returning value
    # w = w * (y_max - y_min)
    # b = b * (y_max - y_min) + y_min
    y_min = min(y_train)
    y_max = max(y_train)
    normalize = lambda y : (y - y_min)/(y_max - y_min)

    ### IMPLEMENT FROM 

    # parameters
    step_num = 10000
    lr = 1e-3
    w, b = 0, 0
    dw, db = 0, 0

    # type-casting
    x = np.array(x_train)
    y = normalize(np.array(y_train))

    for i in range(step_num):
        dw = (w*x + b - y)*x
        db = (w*x + b - y)
        w -= lr * dw.mean()
        b -= lr * db.mean()

        #if (i%1000) == 0:

    w = w * (y_max - y_min)
    b = b * (y_max - y_min) + y_min
    return w, b

    # for i in range(step_num):
    #     pred = torch.mul(w, x_train) + b
    #     loss = pred - y_train
    #     mse = torch.mean(loss*loss)
    #     mse.backward()
    #     w.data -= lr*w.grad.data
    #     b.data -= lr*w.grad.data
    #     w.grad.data.zero_()
    #     b.grad.data.zero_()
    #     if (i%1000==0):
    #     #x.data -= lr*x.grad.data
    
    # return w, b

def predict(x_train : list[float], y_train : list[float], x_test : list[float]) -> list[float]: # DO NOT MODIFY FUNCTION NAME
    ### IMPLEMENT FROM HERE
    w, b = training(x_train, y_train)
    result = np.array(x_test) * w - b
    return list(result)

if __name__ == "__main__":
    x_train = [0.0, 1.0, 2.0, 3.0, 4.0]
    y_train = [2.0, 4.0, 6.0, 8.0, 10.0] # Note that not all test cases give clear line.
    x_test = [5.0, 10.0, 8.0]
    
    w, b = training(x_train, y_train)
    y_test = predict(x_train, y_train, x_test)

    print(w, b)
    print(y_test)
